Update residual line when model line is changed.

HanHsuanWu opened the pull request:
As suggested in #3352, I rebase to RELEASE_next_patch, as well as cleaning up the commits history.

Currently, the plotted residual is not being updated whenever the model line changes.
So I added update for model._residual_line whenever mode._model_line is being updated.
```python
import numpy as np
import hyperspy.api as hs

# Random Signal1D signal
s = hs.signals.Signal1D(data=np.random.random((2, 2, 1024)), 
                        axes=[{'name': 'x', 'size': 2}, 
                              {'name': 'y', 'size': 2}, 
                              {'name': 'sig', 'size': 1024}])
# Model object with Polynomial component
m = s.create_model()
# Polynomial=hs.model.components1D.Polynomial()
# m.append(Polynomial)
m.append(hs.model.components1D.PowerLaw())
m.plot(plot_residual=True)
```

HanHsuanWu committed:
Update residual line when model line is changed.

Update residual line when model line is changed. #3352

ericpre commented on hyperspy/model.py:
```diff
@@ -705,17 +705,29 @@ def _plot_active(self):
             return False
 
-    def _connect_parameters2update_plot(self, components):
+    def _connect_parameters2update_plot(self, components, update_residual=False):
```
```suggestion
    def _connect_parameters2update_plot(self, components):
```

ericpre commented on hyperspy/model.py:
```diff
@@ -705,17 +705,29 @@ def _plot_active(self):
-                parameter.events.value_changed.connect(
-                    self._model_line._auto_update_line, []
-                )
+
+        if update_residual: #update both model and its residual lines
+            for i, component in enumerate(components):
+                component.events.active_changed.connect(
+                    self._model_line._auto_update_line, [])
+                component.events.active_changed.connect(
+                    self._residual_line.update, [])
+                for parameter in component.parameters:
+                    parameter.events.value_changed.connect(
+                        self._model_line._auto_update_line, [])
+                    parameter.events.value_changed.connect(
+                        self._residual_line.update, [])
+
+        else: #update only model line
+            for i, component in enumerate(components):
+                component.events.active_changed.connect(
+                    self._model_line._auto_update_line, [])
+                for parameter in component.parameters:
+                    parameter.events.value_changed.connect(
+                        self._model_line._auto_update_line, [])
```
```suggestion
        lines = [line for line in (self._model_line, self._residual_line) if line is not None]
        for i, component in enumerate(components):
            for line in lines:
                component.events.active_changed.connect(
                    line ._auto_update_line, []
                )
                for parameter in component.parameters:
                    parameter.events.value_changed.connect(
                        line ._auto_update_line, []
                    )
```

ericpre commented on hyperspy/models/model1d.py:
```diff
@@ -685,6 +685,11 @@ def plot(self, plot_components=False, plot_residual=False, **kwargs):
             l3.plot()
             # Quick access to _residual_line if needed
             self._residual_line = l3
+            #Update residual line when model line is updated
+            self._connect_parameters2update_plot(self, update_residual=True) 
+        else:
+            #Update only the model line
+            self._connect_parameters2update_plot(self, update_residual=False)
```
```suggestion
```

ericpre commented on hyperspy/models/model1d.py:
```diff
@@ -672,7 +672,7 @@ def plot(self, plot_components=False, plot_residual=False, **kwargs):
         self._model_line = l2
         self._plot = self.signal._plot
-        self._connect_parameters2update_plot(self)
+
```
```suggestion
        self._connect_parameters2update_plot(self)
```

ericpre reviewed (commented):
I have suggested some simplication to avoid boilerplate code with the idea that if `self._residual_line` is not None, then it should be connected.
The same would need to be done for disconnecting the line, when closing the figure and most likely `self._residual_line` will need to be reset to `None` - I don't think this is the case at the moment.

I didn't test that my suggestion are corrects (only by reading the code), so I may be missing something!

It would need a test to check that event are connected and disconnected properly, for example something along the lines of: https://github.com/hyperspy/hyperspy/blob/3b350ff80b68b0928b284e76d3dc0a2b31e6dbf5/hyperspy/tests/drawing/test_plot_signal_tools.py#L91-L100

HanHsuanWu reviewed (commented):
nice I will test the changes soon

HanHsuanWu committed:
Apply suggestions from code review

Co-authored-by: Eric Prestat <eric.prestat@gmail.com>

HanHsuanWu committed:
Apply suggestions from code review

Co-authored-by: Eric Prestat <eric.prestat@gmail.com>

HanHsuanWu commented:
> I have suggested some simplication to avoid boilerplate code with the idea that if `self._residual_line` is not None, then it should be connected. The same would need to be done for disconnecting the line, when closing the figure and most likely `self._residual_line` will need to be reset to `None` - I don't think this is the case at the moment.
> 
> I didn't test that my suggestion are corrects (only by reading the code), so I may be missing something!
> 
> It would need a test to check that event are connected and disconnected properly, for example something along the lines of:
> 
> https://github.com/hyperspy/hyperspy/blob/3b350ff80b68b0928b284e76d3dc0a2b31e6dbf5/hyperspy/tests/drawing/test_plot_signal_tools.py#L91-L100

I tried running test_plot_BackgroundRemoval_close_figure() and didn't get any error so I think the connection is being closed?
However, I try to use https://github.com/hyperspy/hyperspy/issues/3316#issuecomment-2036732009 ROI to select signal fitting range and this error message comes up after I plot with plot_residual=True. 
```
Traceback (most recent call last):
  File "C:\Users\Han-Hsuan\.conda\envs\hs20\lib\site-packages\matplotlib\cbook\__init__.py", line 287, in process
    func(*args, **kwargs)
  File "C:\Users\Han-Hsuan\Documents\GitHub\hyperspy\hyperspy\external\matplotlib\widgets.py", line 210, in onmove
    self._onmove(event)
  File "C:\Users\Han-Hsuan\Documents\GitHub\hyperspy\hyperspy\external\matplotlib\widgets.py", line 619, in _onmove
    self.onmove_callback(vmin, vmax)
  File "C:\Users\Han-Hsuan\Documents\GitHub\hyperspy\hyperspy\drawing\_widgets\range.py", line 160, in _span_changed
    self.events.changed.trigger(self)
  File "<string>", line 4, in trigger
  File "C:\Users\Han-Hsuan\Documents\GitHub\hyperspy\hyperspy\events.py", line 424, in trigger
    function(**{kwf: kwargs[kwt] for kwt, kwf in kwsd.items()})
  File "C:\Users\Han-Hsuan\Documents\GitHub\hyperspy\hyperspy\roi.py", line 476, in _on_widget_change
    self.events.changed.trigger(self)
  File "<string>", line 4, in trigger
  File "C:\Users\Han-Hsuan\Documents\GitHub\hyperspy\hyperspy\events.py", line 421, in trigger
    function(**{kw: kwargs.get(kw, None) for kw in kwsl})
  File "C:\Users\Han-Hsuan\Documents\GitHub\vibrational_EELS\hs_functions.py", line 498, in _select_signal_range
    hs_model.fit()
  File "C:\Users\Han-Hsuan\Documents\GitHub\hyperspy\hyperspy\model.py", line 1682, in fit
    res = leastsq(
  File "C:\Users\Han-Hsuan\.conda\envs\hs20\lib\site-packages\scipy\optimize\_minpack_py.py", line 426, in leastsq
    retval = _minpack._lmdif(func, x0, args, full_output, ftol, xtol,
  File "C:\Users\Han-Hsuan\Documents\GitHub\hyperspy\hyperspy\models\model1d.py", line 396, in _errfunc
    errfunc = self._model_function(param) - y
  File "C:\Users\Han-Hsuan\Documents\GitHub\hyperspy\hyperspy\model.py", line 1067, in _model_function
    self._fetch_values_from_p0()
  File "C:\Users\Han-Hsuan\Documents\GitHub\hyperspy\hyperspy\model.py", line 1041, in _fetch_values_from_p0
    component.fetch_values_from_array(
  File "C:\Users\Han-Hsuan\Documents\GitHub\hyperspy\hyperspy\component.py", line 1014, in fetch_values_from_array
    parameter.value = p[i] if length == 1 else p[i : i + length]
  File "C:\Users\Han-Hsuan\Documents\GitHub\hyperspy\hyperspy\component.py", line 344, in _set_value
    self.events.value_changed.trigger(value=self.__value, obj=self)
  File "<string>", line 4, in trigger
  File "C:\Users\Han-Hsuan\Documents\GitHub\hyperspy\hyperspy\events.py", line 421, in trigger
    function(**{kw: kwargs.get(kw, None) for kw in kwsl})
  File "C:\Users\Han-Hsuan\Documents\GitHub\hyperspy\hyperspy\drawing\signal1d.py", line 464, in _auto_update_line
    self.update(self, update_ylimits=update_ylimits, **kwargs)
  File "C:\Users\Han-Hsuan\Documents\GitHub\hyperspy\hyperspy\drawing\signal1d.py", line 548, in update
    self.ax.hspy_fig.render_figure()
  File "C:\Users\Han-Hsuan\Documents\GitHub\hyperspy\hyperspy\drawing\figure.py", line 140, in render_figure
    if self.figure.canvas.supports_blit and self._background is not None:
AttributeError: 'NoneType' object has no attribute 'canvas'
```

HanHsuanWu committed:
minor fix

self._connect_parameters2update_plot(self) should be called after plot_residual is created otherwise it will not update

HanHsuanWu committed:
self.residual_line disconnection and reset to None when plot is closed

HanHsuanWu commented:
@ericpre. Took me a while to understand the testing part. But after updating the disconnection, here is a test I made.
```
def test_plot_residual_close_figure(): 
    s = signals.Signal1D(np.arange(1000).reshape(10, 100)) 
    md = s.create_model()
    md.append(hs.model.components1D.Gaussian())
    md.plot(plot_residual=True)
    signal_plot = s._plot.signal_plot

    assert len(s._plot.signal_plot.events.closed.connected) == 5
    assert len(s.axes_manager.events.indices_changed.connected) == 3
    s._plot.close() 
    print("signal_plot.events.closed.connected:",signal_plot.events.closed.connected)
    print("s.axes_manager.events.indices_changed.connected:",s.axes_manager.events.indices_changed.connected)

test_plot_residual_close_figure()
```
Here is the result.
signal_plot.events.closed.connected: set()
s.axes_manager.events.indices_changed.connected: {<bound method BaseModel._on_navigating of <Model1D>>}

I am not sure what "<bound method BaseModel._on_navigating of <Model1D>>" is and how to disconnect that.

ericpre commented:
@HanHsuanWu, I will look at this in a couple of days.

ericpre commented:
This is a function that fetch the current value of the model when the indices are changed and this should stay connected.

It is connected during initialisation of the class:
https://github.com/hyperspy/hyperspy/blob/RELEASE_next_patch/hyperspy/models/model1d.py#L242

and call:
https://github.com/hyperspy/hyperspy/blob/b742845d7f606bc4086f5bec4bc0ca84b8e4104d/hyperspy/model.py#L997-L1001

If you check that the number of connected functions is the same before plotting and after closing the plot, that should be good enough.

ericpre committed:
Add test for checking events connection and resetting plotting line when closing model plot

ericpre committed:
Add changelog entry

ericpre reviewed (approved):
Added a tests and a changelog entry to finish this PR.

HanHsuanWu commented:
Thank you @ericpre for finishing this up.
